src/lib/header-bidding/prebid/modules/analyticsAdapter.ts

```diff
@@ -91,235 +69,255 @@ function logEvents(events: EventData[]): void {
 	log('commercial', `Prebid.js events: ${logMsg}`, events);
 }
 
-// Protect against setting undefined or null values
-function setSafely<T extends Record<string, unknown>, K extends string>(
-	obj: T,
-	key: K,
-	value: unknown,
-): void {
-	if (value === undefined || value === null) {
-		return;
-	}
-	obj = { ...obj, [key]: value };
+function isEventKey(key: string): key is keyof EventData {
+	return eventKeys.includes(key as keyof EventData);
 }
 
-const trackBidWon: Handler = (_, args: BidArgs) => {
-	const event: EventData = { ev: 'bidwon' };
-	setSafely(event, 'aid', args.auctionId);
-	setSafely(event, 'bid', args.requestId);
-	return [event];
-};
-
-const trackAuctionInit: Handler = (adapter, args) => {
-	if (adapter.context) {
-		adapter.context.auctionTimeStart = Date.now();
+// Remove any properties that are undefined or null
+function createEvent(event: RawEventData): EventData {
+	if (!event.ev) {
+		throw new Error('Event must have an "ev" property');
 	}
-	const event: EventData = { ev: 'init' };
-	setSafely(event, 'aid', args.auctionId);
-	setSafely(event, 'st', adapter.context?.auctionTimeStart);
-	return [event];
-};
-
-const trackBidRequest: Handler = (_, args) => {
-	if (args.bids) {
-		return args.bids.map((bid) => {
-			const event: EventData = { ev: 'request' };
-			setSafely(event, 'n', args.bidderCode);
-			setSafely(event, 'sid', bid.adUnitCode);
-			setSafely(event, 'bid', bid.bidId);
-			setSafely(event, 'st', args.start);
-			return event;
-		});
+	const cleanedEvent: Partial<EventData> = {
+		ev: event.ev,
+	};
+	for (const key in event) {
+		if (
+			isEventKey(key) &&
+			event[key] !== undefined &&
+			event[key] !== null
+		) {
+			cleanedEvent[key] = event[key];
+		}
 	}
-	return null;
-};
 
-const trackBidResponse: Handler = (_, args) => {
-	if (args.statusMessage === 'Bid available') {
-		const event: EventData = { ev: 'response' };
-		setSafely(event, 'n', getBidderCode(args));
-		setSafely(event, 'bid', args.requestId);
-		setSafely(event, 'sid', args.adUnitCode);
-		setSafely(event, 'cpm', args.cpm);
-		setSafely(event, 'pb', args.pbCg);
-		setSafely(event, 'cry', args.currency);
-		setSafely(event, 'net', args.netRevenue);
-		setSafely(event, 'did', args.adId);
-		setSafely(event, 'cid', args.creativeId);
-		setSafely(event, 'sz', args.size);
-		setSafely(event, 'ttr', args.timeToRespond);
-		setSafely(event, 'lid', args.dealId);
+	return cleanedEvent as EventData;
+}
+
+// Handlers for each event type, these create a loggable event from prebid event
+const handlers = {
```

Comment:
having the handlers like this gives us a few benefits, we can ensure that the function name matches the event and eliminates the switch statement